src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ValidationHelpers.cs

```diff
@@ -57,50 +58,72 @@ public static string GetModelErrorMessageOrDefault(ModelError modelError)
                 {
                     return new[] { ms };
                 }
-
-                return Enumerable.Empty<ModelStateEntry>();
             }
-            else
+            else if (viewData.ModelState.Count > 0)
             {
                 var metadata = viewData.ModelMetadata;
-                var orderer = new ErrorsOrderer(metadata);
+                var modelStateDictionary = viewData.ModelState;
+                var entries = new List<ModelStateEntry>();
+                Visit(modelStateDictionary, modelStateDictionary.Root, metadata, entries);
+
+                Debug.Assert(entries[entries.Count - 1] == modelStateDictionary.Root);
+                // Special case the root node. It is only 'Visible` if it has any errors associated with it.
+                if (modelStateDictionary.Root.Errors.Count == 0)
+                {
+                    entries.RemoveAt(entries.Count - 1);
+                }
+
+                if (entries.Count < modelStateDictionary.Count)
+                {
+                    // Account for entries in the ModelStateDictionary that do not have corresponding ModelMetadata values.
+                    foreach (var entry in modelStateDictionary)
+                    {
+                        if (!entries.Contains(entry.Value))
+                        {
+                            entries.Add(entry.Value);
+                        }
+                    }
+                }
 
-                return viewData.ModelState
-                    .OrderBy(data => orderer.GetOrder(data.Key))
-                    .Select(ms => ms.Value);
+                return entries;
             }
+
+            return EmptyModelStateEntries;
         }
 
-        // Helper for sorting modelStates to respect the ordering in the metadata.
-        // ModelState doesn't refer to ModelMetadata, but we can correlate via the property name.
-        private class ErrorsOrderer
+        private static void Visit(
+            ModelStateDictionary dictionary,
+            ModelStateEntry modelStateEntry,
+            ModelMetadata metadata,
+            List<ModelStateEntry> orderedModelStateEntries)
         {
-            private readonly Dictionary<string, int> _ordering =
-                new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
-
-            public ErrorsOrderer(ModelMetadata metadata)
+            if (metadata.ElementMetadata != null)
             {
-                if (metadata == null)
+                foreach (var indexEntry in modelStateEntry.Children)
```

Comment:
`Children` can include `!Visible` nodes but that seems fine given invisible nodes won't have `Error`s. If a `ChildKeys` property were exposed as well, could `Visit()` just recurse through the entire tree in one way and in the order it prefers? The only special case might relate to `ElementMetadata` versus `metadata.Properties[subKey]`.